Delete func rework implementing memory using threading.

SpencerMartel opened the pull request:
Delete function has been reworked to allow the bot to suggest titles to delete if there are multiple instances of a game with the same name (ex. delete "Lego" could be asking to delete any number of the top tier lego star wars games). This functions very similarly to how add games works if the user does not input an exact title name.

There has also been a small retouching in terms of formatting. The bot will now tell you the the current price of a game and if it is on sale when you add it to your wishlist. Some other small changes like changing wording from "our tracking list" to "your wishlist" to make the bot seem more personable and as though it is the user in control.

The games_map now gets updated weekly on the same timer as the friday function. It may be worth it to have it update daily though.

I have also changed some of the function names to be clearer in saying what they do.

SpencerMartel committed:
Delete func memory using threading.

Delete function has been reworked to allow the bot to suggest titles to delete if there are multiple instances of a game with the same name (ex. delete "Lego" could be asking to delete any number of the top tier lego star wars games). This functions very similarly to how add games works if the user does not input an exact title name.

There has also been a small retouching in terms of formatting. The bot will now tell you the the current price of a game and if it is on sale when you add it to your wishlist. Some other small changes like changing wording from "our tracking list" to "your wishlist" to make the bot seem more personable and it is the user in control.

SpencerMartel committed:
Clearing up comment

Nwitz commented on client.py:
```diff
@@ -65,9 +68,9 @@ async def on_message(message):
-        # elif command == 'clear':
-        #     clear_wishlist()
-        #     await message.reply ('The wishlist has been cleared.')
+        elif command == 'clear':
+            clear_wishlist()
+            await message.reply ('The wishlist has been cleared.')
```
Do we want this enabled in production?

SpencerMartel replied:
Nope, shall comment out

Nwitz commented on client.py:
```diff
@@ -146,25 +149,71 @@ async def handle_add_game_request(message, game_name):
+    game_tupple = game_matches[index]
+    deleted_game_result = delete_game(game_tupple)
+    print('Deleted game result is:', deleted_game_result)
+    if deleted_game_result == True:
+        reply = f'**{game_tupple[1]}** was successfully deleted from your wishlist.'
+    await message.reply(reply)
```
We are conditionally sending a reply? So if the game doesn't get deleted, it is a silent failure?

SpencerMartel replied:
Yes, currently a silent failure, shall fix in next commit

Nwitz commented on client.py:
```diff
@@ -146,25 +149,71 @@ async def handle_add_game_request(message, game_name):
+    if len(matches) == 1:
+        # Working with a list with a single tupple in it from matches
+        tupple = (matches[0][0], matches[0][1])
+        status = delete_game(tupple)
```
Won't this always delete the game, even if the entered value doesn't match exactly? So the IF-ELSE below will always be true unless some other failure occurs?

SpencerMartel replied:
Yup this deletes the game if only 1 exists with "star wars" in it, sort of assuming that was the correct one. Is this dangerous? In the add we have the "is this the game you meant" prompt but wasn't sure how to implement that for delete (we spoke about having an if for which side the argument came from - add or delete)

Nwitz replied:
Yeah we could add state to the ClientStateManager to keep track of which operation we are storing the list/handling stuff for

SpencerMartel replied:
Think this is something we will have to pair program, not sure how to get it done with the ClientStateManager class

Nwitz commented on client.py:
```diff
@@ -146,25 +149,71 @@ async def handle_add_game_request(message, game_name):
+        else: # Not exact match
+            reply = f'Did you mean **{game_title}**? Write "Yes" to confirm'
+            client_state.store_game_matches(matches)
+    elif len(matches) >1:
```
```suggestion
    elif len(matches) > 1:

```

Nwitz commented on client.py:
```diff
@@ -146,25 +149,71 @@ async def handle_add_game_request(message, game_name):
     else:
-        output = games_were_tracking_string()
-        reply = f'There was a problem deleting the game, are we tracking it?\n{output}'
+        reply = 'This game doesn\'t exist on steam, try gamepass.'
```
Could this be confusing since we're trying to delete?

SpencerMartel replied:
yup, resolved

Nwitz commented on client.py:
```diff
@@ -146,25 +149,71 @@ async def handle_add_game_request(message, game_name):
     await message.reply (reply)
 
+# Formatting function for determining if a game is on sale and giving back a string to add to a reply.
+def price_formatter(queried_key):
+    game_info = get_complete_info_for_game(queried_key)
+    discounted_percent = game_info['price_overview']['discount_percent']
+    game_price = game_info['price_overview']['final_formatted']
+    formatted_game_price = game_price.replace('CDN$ ', '$')
+    if discounted_percent > 0:
+        price_reply = f'and is **currently on sale** for {formatted_game_price} - {discounted_percent}% off!'
+    else:
+        price_reply = f'and is currently full price at {formatted_game_price}.'
+    return price_reply
```
Nice

Nwitz commented on client.py:
```diff
@@ -221,7 +270,7 @@ def games_were_tracking_string():
 def format_game_for_reply(game, game_id):
     url = f'steam://openurl/https://store.steampowered.com/app/{game_id}'
     name = game['name']
-    discounted_percent = game['price_overview']['discount_percent']
+    discounted_percent = game['price_overview']
```
Maybe good time to rename discounted_percent since it is actually holding price_overview I believe

SpencerMartel replied:
I honestly am not sure why I changed this, I think this would actually break it since discounted_percent is later referenced in an f string, so I definitely just want it to have the number not the entire price_overview. I have reverted this.

Nwitz commented on client.py:
```diff
@@ -311,6 +360,7 @@ async def configure_daily_wishlist_check():
 async def friday_reminder():
     channel = client.get_channel(discord_config["channel_id"])
     message = friday_reminder_formatter()
+    fetch_games_mapping()
```
Good call

Nwitz commented on steam.py:
```diff
@@ -37,36 +36,59 @@ def fetch_games_mapping():
     games_list.close
 
-# Read the locally stored apps list retrieved from the 
+# Read all locally stored apps list retrieved from the steam API.
```
Good improvement, this is clearer

Nwitz commented on steam.py:
```diff
@@ -37,36 +36,59 @@ def fetch_games_mapping():
     f.close()
     return json.loads(games_list)
 
+# Read locally stored wishlist file.
+def read_wishlist_json():
```
About time we have some good naming 👍

Nwitz commented on steam.py:
```diff
@@ -217,14 +247,20 @@ def friday_phrase_randomizer():
+    return dict_info
+
 def clear_special_characters(string_to_manulate):
+    special_characters = [u"\u2122", u"\u00AE"]
```
Keep in mind these are characters we may want to be aware of in other places, for now since we only use it here it makes sense to keep it local to the function, but might want to move in the future, maybe when we create a config file or something similar

Nwitz reviewed (commented):
Great work on this - big change with complexity

SpencerMartel committed:
Update client.py

Co-authored-by: Noah Horowitz <Foolzo13@gmail.com>

SpencerMartel committed:
Working Through comments

Working through your comments in this commit.
1. Clear is now commented out.
2. We no longer throw a silent error if the new delete function does not work.
3. Re-added a field that I believe was mistakenly deleted.

SpencerMartel committed:
Merge remote-tracking branch 'origin/Delete_rework' into Delete_rework
